lib/FederationService.php

```diff
@@ -0,0 +1,215 @@
+ *
+ */
+namespace OCA\Richdocuments;
+
+use OCP\ILogger;
+use OCP\Http\Client\IClientService;
+use OCP\IURLGenerator;
+
+class FederationService {
+	/**
+	 * @var ILogger
+	 */
+	private $logger;
+
+	/**
+	 * @var IURLGenerator
+	 */
+	private $urlGenerator;
+
+	/**
+	 * @var IClientService
+	 */
+	private $httpClient;
+
+	public function __construct(
+		ILogger $logger,
+		IURLGenerator $urlGenerator,
+		IClientService $httpClient
+	) {
+		$this->logger = $logger;
+		$this->urlGenerator = $urlGenerator;
+		$this->httpClient = $httpClient;
+	}
+
+	/**
+	 * Get the Url of the collabora document on a federated server.
+	 *
+	 * @param string $shareToken
+	 * @param string $shareRelativePath
+	 * @param string $server
+	 * @param string $accessToken
+	 * @return string with the Url to the given resource
+	 */
+	public function getRemoteFileUrl(string $shareToken, string $shareRelativePath, string $server, string $accessToken) : string {
+		$serverHost = $this->urlGenerator->getAbsoluteURL('/');
+		$remoteFileUrl = \rtrim($server, '/') . '/index.php/apps/richdocuments/documents.php/federated' .
+			'?shareToken=' . $shareToken .
+			'&shareRelativePath=' . $shareRelativePath .
+			'&server=' . \rtrim($serverHost, '/') .
+			'&accessToken=' . $accessToken;
+		return $remoteFileUrl;
+	}
+	
+	/**
+	*
+	* @param string $server addres of a remote server
+	* @param string $accessToken wopi access token from a remote server
+	* @return array|null with additional wopi information
+	*/
+	public function getWopiForToken($server, $accessToken) {
+		$remote = $server;
+
+		if (!$this->isServerAllowed($remote)) {
+			$this->logger->info("Server {server} is not allowed.", ["server" => $remote]);
+			return null;
+		}
+
+		try {
+			$client = $this->httpClient->newClient();
+			$url = $remote . '/ocs/v2.php/apps/richdocuments/api/v1/federation';
+
+			$response = $client->post(
+				$url,
+				[
+					'form_params' => [
+						'token' => $accessToken,
+						'format' => 'json'
+					],
+					'timeout' => 3,
+					'connect_timeout' => 3,
+				]
+			);
+
+			$responseBody = $response->getBody();
+			$data = \json_decode($responseBody, true, 512);
+			if (\is_array($data)) {
+				return $data['ocs']['data'];
+			}
+			return null;
+		} catch (\Throwable $e) {
+			$this->logger->info('Cannot get the wopi info from remote server: ' . $remote, ['exception' => $e]);
+		}
+
+		return null;
+	}
+
+	/**
+	 * Check if server is allowed
+	 *
+	 * @param string $remote a remote url
+	 * @return bool indicating if given remote is allowed server
+	 */
+	public function isServerAllowed($remote) {
+		// TODO: implement check for trusted server, for a moment all trusted
+
+		return true;
+	}
+
+	/**
+	 * Get the wopiSrc Url from a remote server.
+	 *
+	 * @param string $server a remote
+	 * @return string with the wopi src Url
+	 */
+	public function getRemoteWopiSrc($server) {
+		if (\strpos($server, 'http://') === false && \strpos($server, 'https://') === false) {
```

Comment:
I'm not sure if it's possible, but if server is "abcd/efghttp://my.server.com" we won't be prefixing the protocol.
If we expect the server to have a specific format, for example, to be urlencoded, then it should be at least specified in the PHPDoc, otherwise I assume any string is valid.